src/r_sky.c

```diff
@@ -42,255 +44,365 @@ boolean stretchsky;
+// [EA] done away with original spreadFire, in favor of Odamex's texture-
+// replacing implementation.
+static void SpreadFire(int src, byte *fire, int width)
 {
-    byte *rover = fire_pixels;
-    for (int x = 0; x < FIRE_WIDTH; x++)
+    const byte pixel = fire[src];
+    const int copyloc0 = src - width;
+    if (pixel == 0)
     {
-        byte *src = fire_indices + x;
-        for (int y = 0; y < FIRE_HEIGHT; y++)
+        if (copyloc0 >= 0)
         {
-            *rover++ = fire->palette[*src];
-            src += FIRE_WIDTH;
+            fire[copyloc0] = 0;
+        }
+    }
+    else
+    {
+        const int rand = M_Random() & 3;
+
+        const int copyloc1 = copyloc0 - rand + 1;
+        if (copyloc1 >= 0)
+        {
+            fire[copyloc1] = pixel - (rand & 1);
         }
     }
 }
 
-static void SpreadFire(void)
+static void UpdateFireSky(sky_t *sky)
 {
-    for (int x = 0; x < FIRE_WIDTH; ++x)
+    // [EA] [RF]
+    // the fire algorithm affects multiple collums at once, therefore both XY
+    // loops _must_ be separated, the following will cause noticeable visual
+    // disturbance on the resulting fire effect
+
+    const int texnum = sky->background.texture;
+    int height = textures[texnum]->height;
+    int width = textures[texnum]->width;
+    byte *coldata;
+
+    for (int x = 0; x < width; x++)
     {
-        for (int y = 1; y < FIRE_HEIGHT; ++y)
+        for (int y = 1; y < height; y++)
         {
-            int src = y * FIRE_WIDTH + x;
-
-            int index = fire_indices[src];
+            int src = y * width + x;
+            SpreadFire(src, sky->fire, width);
+        }
+    }
 
-            if (!index)
-            {
-                fire_indices[src - FIRE_WIDTH] = 0;
-            }
-            else
-            {
-                int rand_index = M_Random() & 3;
-                int dst = src - rand_index + 1;
-                fire_indices[dst - FIRE_WIDTH] = index - (rand_index & 1);
-            }
+    for (int x = 0; x < width; x++)
+    {
+        coldata = R_GetColumn(texnum, x);
+        for (int y = 0; y < height; y++)
+        {
+            int src = y * width + x;
+            coldata[y] = sky->palette[sky->fire[src]];
         }
     }
 }
 
-static void SetupFire(fire_t *fire)
+void R_InitFireSky(sky_t *sky)
 {
-    memset(fire_indices, 0, FIRE_WIDTH * FIRE_HEIGHT);
-
-    int last = array_size(fire->palette) - 1;
+    int texnum = sky->background.texture;
+    const texture_t *tex = textures[texnum];
+    size_t size = tex->width * tex->height;
+    sky->fire = Z_Calloc(1, size, PU_STATIC, NULL);
 
-    for (int i = 0; i < FIRE_WIDTH; ++i)
+    for (int i = 0; i < tex->width * tex->height; i++)
     {
-        fire_indices[(FIRE_HEIGHT - 1) * FIRE_WIDTH + i] = last;
+        sky->fire[i] = 0;
     }
 
-    for (int i = 0; i < 64; ++i)
+    for (int i = 0; i < tex->width; i++)
     {
-        SpreadFire();
+        sky->fire[(tex->height - 1) * tex->width + i] = sky->palette_num - 1;
```

Comment:
We don't need `palette_num` field in `sky_t`, we can replace it with `array_size(sky->palette)`.